src/commands.cpp

```diff
@@ -7074,6 +7092,266 @@ void CommandSetLastAcknowledged::procresult()
+
+    if (isverificationcode(verificationcode))
+    {
+        arg("c", verificationcode.c_str());
+    }
+
+    tag = client->reqtag;
+}
+
+bool CommandSMSVerificationCheck::isverificationcode(const string& s)
+{
+    for (const char c : s)
+    {
+        if (!isdigit(c))
+        {
+            return false;
+        }
+    }
+    return s.size() == 6;
+}
+
+void CommandSMSVerificationCheck::procresult()
+{
+    if (client->json.isnumeric())
+    {
+        client->app->smsverificationcheck_result(static_cast<error>(client->json.getint()));
+    }
+    else
+    {
+        client->json.storeobject();
+        client->app->smsverificationcheck_result(API_EINTERNAL);
+    }
+}
+
+CommandGetRegisteredContacts::CommandGetRegisteredContacts(MegaClient* client, const map<const char*, const char*>& contacts)
+{
+    cmd("usabd");
+
+    tag = client->reqtag;
+
+    beginobject("e");
+    for (const auto& pair : contacts)
+    {
+        arg(pair.first, pair.second);
+    }
+    endobject();
+}
+
+void CommandGetRegisteredContacts::procresult()
+{
+    processResult(*client->app, client->json);
+}
+
+void CommandGetRegisteredContacts::processResult(MegaApp& app, JSON& json)
+{
+    if (json.isnumeric())
+    {
+        app.getregisteredcontacts_result(static_cast<error>(json.getint()), nullptr);
+        return;
+    }
+
+    vector<tuple<string, string, string>> registered_contacts;
+
+    string entry_user_detail;
+    string id;
+    string user_detail;
+
+    for (;;)
+    {
+        if (json.enterobject())
+        {
+            bool end_of_object = false;
+            while (!end_of_object)
+            {
+                switch (json.getnameid())
+                {
+                    case MAKENAMEID3('e', 'u', 'd'):
+                    {
+                        json.storeobject(&entry_user_detail);
+                        break;
+                    }
+                    case MAKENAMEID2('i', 'd'):
+                    {
+                        json.storeobject(&id);
+                        break;
+                    }
+                    case MAKENAMEID2('u', 'd'):
+                    {
+                        json.storeobject(&user_detail);
+                        break;
+                    }
+                    case EOO:
+                    {
+                        end_of_object = true;
+                        break;
+                    }
+                    default:
+                    {
+                        if (!json.storeobject())
+                        {
+                            LOG_err << "Failed to parse 'get registered contacts' response";
+                            app.getregisteredcontacts_result(API_EINTERNAL, nullptr);
+                            return;
+                        }
+                    }
+                }
+            }
+            json.leaveobject();
+            if (entry_user_detail.empty() || id.empty() || user_detail.empty())
```

Comment:
I think failing to read one of the items should fail the whole json parsing cuz it means there's something inherently wrong and the data cannot be trusted.

Reply:
The command request and responses are in batches, so the reply from the server is often one big JSON array of individual responses, eg  [{"a":"b"},-9,[1,2,3]].  So we must leave the json pointer pointing to the next response in the overall array, even if this current one can't be parsed (often just due to a newly added field, or a value turns out to be an array instead of a string etc.  That way the next command response can be processed correctly, minimizing the possibility of a new feature breaking existing functionality.

Reply:
Just had a chat with Matt about the json parsing. I will change the parsing to ensure we get to the end of the json in both success and error case. That way a failure of the current json doesn't fail the whole batch.